### PR TITLE
Test the type promotion between every two dtypes thoroughly

### LINKED ISSUE #41842: Test on the correctness of type promotion between each two dtypes
Currently `test/test_type_promotion.py` doesn't seem to have such a comprehensive test. It may be a good idea to add it. Derived from discussion at https://github.com/pytorch/pytorch/pull/41698#discussion_r458463593

cc @mruberry @VitalyFedyunin

### PR DESCRIPTION
**Pull request by xuhdev**
Fixes #41842

### CODE REVIEW

File: test/test_type_promotion.py [hunk 1]
```diff
@@ -345,17 +345,42 @@ def test_create_bool_tensors(self, device):
...
+            # int scalar
+            if dtype == torch.bool:
+                self.assertEqual(torch.result_type(torch.tensor([0, 1], dtype=dtype, device=device), 1), torch.int64)
+                self.assertEqual(torch.result_type(1, torch.tensor([0, 1], dtype=dtype, device=device)), torch.int64)
+            else:
+                self.assertEqual(torch.result_type(torch.tensor([0, 1], dtype=dtype, device=device), 1), dtype)
+                self.assertEqual(torch.result_type(1, torch.tensor([0, 1], dtype=dtype, device=device)), dtype)
+
+            # float scalar
+            if torch.testing.is_integral(dtype):
```
Comment (mruberry) [inline test/test_type_promotion.py:372]: These cases look good. How about adding a complex scalar case, too?

File: test/test_type_promotion.py [hunk 2]
```diff
@@ -345,17 +345,42 @@ def test_create_bool_tensors(self, device):
 
     @float_double_default_dtype
     def test_result_type(self, device):
-        self.assertEqual(torch.result_type(torch.tensor(1, dtype=torch.int, device=device), 1), torch.int)
-        self.assertEqual(torch.result_type(1, torch.tensor(1, dtype=torch.int, device=device)), torch.int)
+        # Tensor against tensor
+        for dtypea in torch.testing.get_all_dtypes():
```
Comment (mruberry) [inline test/test_type_promotion.py:349]: You can simplify this test by using the `@dtypes` decorator like this:

```
@dtypes(*product(torch.testing.get_all_dtypes(), torch.testing.get_all_dtypes())
def test_result_type(self, device, dtypes):
  t = torch.tensor((0, 1), device=device, dtype=dtypes[1])
  t_scalar = torch.tensor(2, device=device, dtype=dtypes[1])
  scalar = b_scalar_tensor.item()
  args = (t, t_scalar, scalar)
  ...
```

Then I think you can simplify the body like this:

```
for a, b in product(args, args):
  try:
    result = a + b
  except RuntimeError:
    with self.assertRaises(RuntimeError):
      torch.promote_types(a, b)
    with self.assertRaises(RuntimeError):
      torch.result_type(a, b)
    continue

  self.assertEqual(result.dtype, torch.promote_types(a, b))
  self.assertEqual(result.dtype, torch.result_type(a, b)) 
```

I think something like that will ensure that `torch.promote_types`, `torch.result_type` and addition are consistent for scalars, tensors with zero dimensions, and tensors with non-zero dimensions.

What are your thoughts?

File: test/test_type_promotion.py (hunk 1 above)
Comment (mruberry) [reply test/test_type_promotion.py:372]: I made a suggestion above that should naturally test for complex scalars, too, so you wouldn't have to add a special case.

File: test/test_type_promotion.py (hunk 2 above)
Comment (xuhdev) [reply test/test_type_promotion.py:349]: 1. I think the scalar part has to be singled out (as what has been done in the current code), because specifying single-element tensor dtype doesn't always take effects (e.g., uint8 would be seen as int64). See https://github.com/pytorch/pytorch/issues/42473#issuecomment-668792855 . Hence I think using `@dtypes` actually complicates the test code. What do you think?

2. I agree that having `+` in the test is a good idea.

File: test/test_type_promotion.py (hunk 1 above)
Comment (xuhdev) [reply test/test_type_promotion.py:372]: Please see my other comment and I'll address this accordingly.

File: test/test_type_promotion.py (hunk 2 above)
Comment (mruberry) [reply test/test_type_promotion.py:349]: The scalar x scalar case is uninteresting, but doesn't result_type work if you have tensor x scalar? You could skip the promote_types check in that case (since it needs a.dtype, unlike the sketch I wrote above).

Comment (xuhdev) [reply test/test_type_promotion.py:349]: What I meant is that the simplification of the code would actually make it more complicated in this situation, because all single-element tensors are seen as scalars, and scalars only have a handful dtypes available. The code would have a lot of branches that singles these cases out, which would be more difficult to read than the current flow.

Comment (mruberry) [reply test/test_type_promotion.py:349]: I think it would be more complicated if you're verifying the dtype is correct like you are now, but it's OK just to verify the dtype is consistent between add, promote_types, and result_types.

Comment (xuhdev) [reply test/test_type_promotion.py]: I updated the code according to your code framework. I think the code does look better, but is not as simple as you might have thought. As the comment that I referred to above, the behavior is different for scalars (single-element tensors are treated as scalars), and has to be singled out when testing and thus adds complications because we are now doing all tests in a single loop. For example, the following two expressions give different results and this is intended:

```
In [11]: torch.tensor([190, 220], dtype=torch.int16) + torch.tensor(190, dtype=torch.int32)
Out[11]: tensor([380, 410], dtype=torch.int16)

In [12]: torch.tensor([190, 220], dtype=torch.int16) + torch.tensor([190, 220], dtype=torch.int32)
Out[12]: tensor([380, 440], dtype=torch.int32)
```

Comment (mruberry) [reply test/test_type_promotion.py]: I was thinking of a test like this:

```
    @dtypes(*itertools.product(torch.testing.get_all_dtypes(), torch.testing.get_all_dtypes()))
    def test_result_type(self, device, dtypes):
        t = torch.tensor((0, 1), device=device, dtype=dtypes[1])
        t_scalar = torch.tensor(2, device=device, dtype=dtypes[1])
        scalar = t_scalar.item()
        args = (t, t_scalar, scalar)

        for a, b in itertools.product(args, args):
            try:
                result = a + b
            except RuntimeError:
                with self.assertRaises(RuntimeError):
                    torch.promote_types(a, b)
                with self.assertRaises(RuntimeError):
                    torch.result_type(a, b)
                continue

        # Skips analyzing the scalar x scalar case
        if not isinstance(result, torch.Tensor):
            return

        # NOTE: promote_types assumes both arguments are tensors with 1+ dims
        if isinstance(a, torch.Tensor) and a.dim() != 0 and \
           isinstance(b, torch.Tensor) and b.dim() != 0:
            self.assertEqual(result.dtype, torch.promote_types(a.dtype, b.dtype))
        self.assertEqual(result.dtype, torch.result_type(a, b))
```

which avoids the additional complexity by skipping the separate promote types impl. What would you think of that?

Comment (xuhdev) [reply test/test_type_promotion.py]: But wouldn't we miss the test on the resulting dtype of a tensor against a single-element tensor or a scalar?

Comment (mruberry) [reply test/test_type_promotion.py]: Yes and no. I think there's a trade-off we can discuss here.

This version doesn't test that `torch.promote_types` is consistent with the result dtype when a scalar or zero-dim tensor is involved in the computation, true, but we also don't expect `torch.promote_types` to always return a result consistent with the result type in those cases. This test does, however, validate that `torch.result_type` is always consistent with the addition. 

The test also doesn't attempt to validate using any "ground truth". It simply tests for consistency. That is, adding two torch.float32 tensors might return a long tensor, and the test wouldn't complain if `torch.promote_types` and `torch.result_type` agreed.

Validating that these functions are correct and not just consistent requires reimplementing, in part or whole, PyTorch's type promotion system in Python. This is tricky, as you point out, but I'm also not sure if it's worth doing. 

What are your thoughts?

Comment (xuhdev) [reply test/test_type_promotion.py]: I think it's probably better to have the current additional tests preserved. This test is on `result_type`, not `+`, and I think it makes sense to ensure that with scalars, it should also produce expected results. What do you think?

Comment (mruberry) [reply test/test_type_promotion.py]: Wait I'm confused, the above does validate the consistency of `torch.result_type` and the dtype produced by addition when scalars and zero-dim tensors are involved. Can you elaborate on what you're thinking?

Comment (xuhdev) [reply test/test_type_promotion.py]: Because this would go backward: The `+` operator is supposed to be consistent with `result_type` (not the other way around) and applies `result_type` (see https://github.com/pytorch/pytorch/blob/64a7939ee56fdc0a34b3de66aa9e6f3c96ef6860/aten/src/ATen/native/BinaryOps.cpp#L60 , https://github.com/pytorch/pytorch/blob/64a7939ee56fdc0a34b3de66aa9e6f3c96ef6860/aten/src/ATen/native/TensorIterator.cpp#L696-L697, and https://github.com/pytorch/pytorch/blob/64a7939ee56fdc0a34b3de66aa9e6f3c96ef6860/aten/src/ATen/native/TensorIterator.cpp#L89 ). Here we would like to verify that `result_type` gives the correct resulting dtype when it should: follow `promote_types` when both are tensors, and follow the special rules when one is tensor and the other is scalar.

Comment (mruberry) [reply test/test_type_promotion.py]: Right. But how does the test I presented fail to do this? I feel like I'm missing something here. It always checks `torch.result_type() == (a+b).dtype`, including in tensor + scalar cases, and it checks against `promote_types` when both inputs are 1D+ tensors, right?

Comment (mruberry) [reply test/test_type_promotion.py]: (We can also Slack about this if that would help.)

Comment (xuhdev) [reply test/test_type_promotion.py]: The issue is that, for tensor + scalar, only `torch.result_type() == (a+b).dtype` is examined, but the correctness of `torch.result_type()` is not examined (this is a test case for `torch.result_type()`, not `+`). Ping me on Slack you still have doubts.

Comment (xuhdev) [reply test/test_type_promotion.py]: As discussed offline, I have updated the test to keep the loop for tensor vs tensor and scalar vs scalar, and keep tensor vs scalar as spot check.

File: test/test_type_promotion.py [hunk 3]
```diff
@@ -343,19 +343,69 @@ def test_create_bool_tensors(self, device):
...
+                raise AssertionError(f"Unkonwn type {x}")
+
+        # tensor against tensor
+        a_tensor = torch.tensor((0, 1), device=device, dtype=dtypes[0])
+        a_single_tensor = torch.tensor(1, device=device, dtype=dtypes[0])
+        a_scalar = a_single_tensor.item()
+        b_tensor = torch.tensor((1, 0), device=device, dtype=dtypes[1])
+        b_single_tensor = torch.tensor(1, device=device, dtype=dtypes[1])
+        b_scalar = b_single_tensor.item()
+        for a, b in ((a_tensor, b_tensor), (a_single_tensor, b_single_tensor), (a_scalar, b_scalar)):
```
Comment (mruberry) [inline test/test_type_promotion.py:372]: This is only testing tensor x tensor, zero dim tensor x zero dim tensor, scalar x scalar, but we want to test the cross of these, like tensor x scalar.

File: test/test_type_promotion.py [hunk 4]
```diff
@@ -343,19 +343,69 @@ def test_create_bool_tensors(self, device):
...
+            else:
+                dtype_res = _get_dtype(result)
+                if a is a_scalar and b is b_scalar and dtype_a == torch.bool and dtype_b == torch.bool:
+                    # special case: in Python, True + True is an integer
+                    self.assertEqual(dtype_res, torch.int64, f"a == {a}, b == {b}")
+                else:
+                    self.assertEqual(dtype_res, torch.result_type(a, b), f"a == {a}, b == {b}")
+                if a is a_scalar and b is b_scalar:  # Python internal type determination is good enough in this case
+                    continue
+                self.assertEqual(dtype_res, torch.promote_types(dtype_a, dtype_b), f"a == {a}, b == {b}")
```
Comment (mruberry) [inline test/test_type_promotion.py:391]: I don't think this assertion will be true if this test includes tensor x scalar tests.

File: test/test_type_promotion.py [hunk 5]
```diff
@@ -343,19 +343,69 @@ def test_create_bool_tensors(self, device):
...
+            try:
+                result = a + b
+            except RuntimeError:
+                with self.assertRaises(RuntimeError):
+                    torch.promote_types(dtype_a, dtype_b)
+                with self.assertRaises(RuntimeError):
+                    torch.result_type(a, b)
+            else:
+                dtype_res = _get_dtype(result)
+                if a is a_scalar and b is b_scalar and dtype_a == torch.bool and dtype_b == torch.bool:
```
Comment (mruberry) [inline test/test_type_promotion.py:384]: Cool special case.

File: test/test_type_promotion.py (hunk 4 above)
Comment (mruberry) [reply test/test_type_promotion.py:391]: One option is to only test `torch.promote_types` when both arguments are 1D+ tensors. Then this wouldn't need `_get_dtype`, either. (The test could also validate that `torch.promote_types` only throws an error if both arguments were 1D+ tensors, too.)

File: test/test_type_promotion.py [hunk 6]
```diff
@@ -343,19 +343,69 @@ def test_create_bool_tensors(self, device):
...
-                         torch.tensor([1, 1], dtype=torch.int, device=device)),
-                         torch.int)
-        self.assertEqual(torch.result_type(torch.tensor([1., 1.], dtype=torch.float, device=device), 1.), torch.float)
-        self.assertEqual(torch.result_type(torch.tensor(1., dtype=torch.float, device=device),
-                         torch.tensor(1, dtype=torch.double, device=device)),
-                         torch.double)
+    def test_result_type_tensor_vs_scalar(self, device):
+        "Spot check some result type for tensor against scalar (including single-element tensor)."
+
+        def _test_spot(a, b, res_dtype):
```
Comment (mruberry) [inline test/test_type_promotion.py:397]: Nice.

File: test/test_type_promotion.py [hunk 7]
```diff
@@ -343,19 +343,69 @@ def test_create_bool_tensors(self, device):
...
-        self.assertEqual(torch.result_type(torch.tensor(1, device=device), 1.), torch.get_default_dtype())
-        self.assertEqual(torch.result_type(torch.tensor(1, dtype=torch.long, device=device),
-                         torch.tensor([1, 1], dtype=torch.int, device=device)),
-                         torch.int)
-        self.assertEqual(torch.result_type(torch.tensor([1., 1.], dtype=torch.float, device=device), 1.), torch.float)
-        self.assertEqual(torch.result_type(torch.tensor(1., dtype=torch.float, device=device),
-                         torch.tensor(1, dtype=torch.double, device=device)),
-                         torch.double)
+    def test_result_type_tensor_vs_scalar(self, device):
+        "Spot check some result type for tensor against scalar (including single-element tensor)."
```
Comment (mruberry) [inline test/test_type_promotion.py:395]: Just make this a comment above the test.

File: test/test_type_promotion.py [hunk 8]
```diff
@@ -343,19 +343,69 @@ def test_create_bool_tensors(self, device):
...
+
+        def _test_spot(a, b, res_dtype):
+            self.assertEqual(torch.result_type(a, b), res_dtype)
+            self.assertEqual(torch.result_type(b, a), res_dtype)
+
+        _test_spot(torch.tensor(1, dtype=torch.int, device=device), 1, torch.int)
```
Comment (mruberry) [inline test/test_type_promotion.py:401]: Add two 1D+ tensor cases:
- half x long -> half
- float x double -> double

And one 0D tensor case:
- complex64 x complex128 -> complex128

File: test/test_type_promotion.py (hunk 7 above)
Comment (xuhdev) [reply test/test_type_promotion.py:395]: Why is that?

File: test/test_type_promotion.py (hunk 3 above)
Comment (xuhdev) [reply test/test_type_promotion.py:372]: tensor x scalar is in the spot checking tests, because it doesn't follow the usual rule followed by `promote_types`.

File: test/test_type_promotion.py (hunk 7 above)
Comment (mruberry) [reply test/test_type_promotion.py:395]: Tests don't typically have docstrings because we don't generate docs for them, that's all. It's just more consistent with our other tests.

File: test/test_type_promotion.py (hunk 3 above)
Comment (mruberry) [reply test/test_type_promotion.py:372]: That's right. See comment below about skipping `promote_types` in those cases. `result_type` should work on them, however.

File: test/test_type_promotion.py (hunk 8 above)
Comment (xuhdev) [reply test/test_type_promotion.py:406]: I added a few spot checks. Since 1D+ x 1D+ are already tested in the test above, I believe you actually meant 1D+ x 0D.

File: test/test_type_promotion.py (hunk 4 above)
Comment (xuhdev) [reply test/test_type_promotion.py]: I have enclosed this line in a very simple `if` block. It would not decrease the coverage of this line at all, but also doesn't add complications.

**Review by mruberry — APPROVED** [review]
Cool!

**Comment by xuhdev** [comment]
Thanks for the review!
